ops/cli.py

```diff
@@ -1,351 +1,343 @@
-    BLACK = 30
-    DEFAULT = 39
+
+class Config:
+    """Base class for storing/resolving/validating command line input."""
+
+    def __init__(self, parsed_args: argparse.Namespace):
+        """Resolve and store input from parsed arguments."""
+        self.command: str = parsed_args.command
+        self.dry_run: bool = parsed_args.dry_run
+        self.force: bool = parsed_args.force
+        self.environment: str = parsed_args.environment
+        info(f"=>  Resolving {self.command} configuration")
+        self.version: str = \
+            self.__resolve_config_value("Version", parsed_args.version, lambda: self.__read_version())
+        self.gotc_deploy_folder: str = \
+            self.__resolve_config_value("GotC Deploy folder", f"gotc-deploy/deploy/{self.environment}")
+        self.instance_id: str = \
+            self.__resolve_config_value("Instance ID", parsed_args.instance)
+        self.project: str = \
+            self.__resolve_config_value("GCP project", parsed_args.project, lambda: f"broad-{self.environment}")
+        self.cloudsql_name: str = \
+            self.__resolve_config_value("CloudSQL name", parsed_args.cloud_sql, lambda: self.__find_cloudsql_name())
+        self.cluster_name: str = \
+            self.__resolve_config_value("GKE cluster name", parsed_args.cluster, lambda: self.__find_cluster_name())
+        self.cluster_zone: str = \
+            self.__resolve_config_value("GKE cluster zone", parsed_args.zone, lambda: self.__find_cluster_zone())
+        self.cluster_namespace: str = \
+            self.__resolve_config_value("Kubernetes namespace", parsed_args.namespace,
+                                        lambda: f"{self.instance_id}-wfl")
+        success("Resolved configuration") if not self.force else warn("Resolved forced configuration")
+
+    @staticmethod
+    def __resolve_config_value(description: str, specific: str, default: Callable[[], str] = lambda: None) -> str:
+        """Either use the specific value if truthy or calculate the default, printing accordingly."""
+        val = specific or default()
+        info(f"    {description.ljust(20, '.')}: {val}{' (specifically supplied)' if specific else ''}")
+        return val
+
+    @staticmethod
+    def __read_version() -> str:
+        with open("version") as version_file:
+            return version_file.read().strip()
+
+    def __find_cloudsql_name(self) -> str:
+        """Use labels to identify the Cloud SQL name."""
+        instances = json.loads(shell(f"gcloud --project {self.project} --format=json "
+                                     "sql instances list "
+                                     f"--filter='labels.app_name=wfl AND labels.instance_id={self.instance_id}'",
+                                     quiet=True))
+        if len(instances) == 1:
+            return instances[0]["name"]
+        else:
+            error("No CloudSQL instance could be inferred")
+            info("    Couldn't find a single match for "
```

Comment:
just throw an exception with all these messages? 

pseudo-python
```python
if __name__ == "__main__":
    try:
        main()
        exit(0)
     except exception:
         print_message( exception ) 
         exit(1)
```

Reply:
I think printing my own messages gives better output. Exceptions certainly still happen, if the shell commands actually run into errors those get passed up. In cases where the command worked and it is a human issue, I can at least say that.

Reply:
Or you can just put the error message into the exception and print that message?

Reply:
It is a discussion of
```
error("my message")
info("my new line")
exit(1)
```
versus
```
raise RuntimeError("my message\nmy new line")
```

I don't think that using exceptions for control flow grants us much benefit in terms of keeping the code clear or easy to maintain, especially when we consider cases where we don't actually want to exit all the way in the case of `--force`. If we had less console output and were comfortable just letting the raw CallProcessExceptions from shell commands rise up to the user, then we can absolutely get rid of all the error text and do that. I firmly believe that helpful error text will save us time and that we shouldn't make code worse in the pursuit of making it only marginally shorter.

Reply:
Resolved in the new version, eliminated `--force` and most error messages to instead use exceptions to stop program flow.